### ISSUE TITLE
torch_cuda link-lib request stripped by linker; LD_PRELOAD required at runtime to enable CUDA

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#9]
## Summary

Building thrust with `LIBTORCH_USE_PYTORCH=1` against a pip-installed PyTorch causes the resulting binary to silently run on CPU even when `torch.cuda.is_available()` is true and the build script emitted `cargo:rustc-link-lib=torch_cuda`. The CUDA backend never initializes, so `Device::cuda_if_available()` returns `Cpu`.

`LD_PRELOAD`'ing `libtorch_cuda.so` and `libc10_cuda.so` at run time forces initialization and the binary then reports `Cuda(0)`. But this is a foot-gun and the failure mode (silent CPU fallback) is hard to notice in benchmarks.

## Repro

On a Linux box with PyTorch >= 2.9 installed via pip and a working NVIDIA driver:

```bash
export LIBTORCH_USE_PYTORCH=1
export LIBTORCH_BYPASS_VERSION_CHECK=1
cargo build --release --example train_p3 --features "training env-bucket-brigade"
ldd target/release/examples/train_p3 | grep torch_cuda  # nothing
```

The build script's `output` file says `cargo:rustc-link-lib=torch_cuda`, but `ldd` shows only `libtorch_cpu.so` and `libc10.so` in the runtime deps. This is the linker's `--as-needed` behavior stripping libs that have no referenced symbols at link time.

## Workaround

```bash
TORCH_LIB=$(python3 -c "import torch, os; print(os.path.join(os.path.dirname(torch.__file__), 'lib'))")
export LD_PRELOAD="$TORCH_LIB/libtorch_cuda.so:$TORCH_LIB/libc10_cuda.so"
./target/release/examples/train_p3 ...
```

After this, `device=Cuda(0)` and training is in fact on the GPU.

## Suggested fixes

1. Have the torch-sys build script emit `cargo:rustc-link-arg=-Wl,--no-as-needed` (or `-Wl,--no-as-needed -ltorch_cuda -Wl,--as-needed` pattern) when CUDA is available, so the linker keeps the unreferenced `libtorch_cuda` dependency.
2. Or, have `tch` itself reference at least one symbol from `libtorch_cuda` (e.g. `at::cuda::is_available()`) so `--as-needed` doesn't strip it.
3. Or, document the `LD_PRELOAD` workaround prominently in `GPU_SETUP.md` so users don't silently fall back to CPU.

## Impact

Discovered while benchmarking thrust against the Python pipeline. Without this fix, the touted "3-6x speedup over Python" can collapse to roughly 1x (CPU vs CPU) because the Rust binary never actually uses the GPU.

## Related

* #4 (multi_agent PolicyLearner TODOs)
* #8 (libtorch dylib mismatch on macOS) --- same family of "tch is brittle against the libtorch we have"

---

### Root Cause Analysis

This bug is a confluence of three behaviors:

1. **`torch-sys` build script** correctly emits `cargo:rustc-link-lib=torch_cuda` when it detects a CUDA-enabled PyTorch (verified in our setup via `LIBTORCH_USE_PYTORCH=1` + cu128 wheel).
2. **GNU `ld` defaults to `--as-needed`** on modern Linux distros (Ubuntu 22.04+, Debian 12+, etc.). Under this flag, the linker drops `DT_NEEDED` entries for shared libraries from which no symbols are referenced. Because `thrust` only calls `tch` high-level APIs and `tch` itself only references `libtorch_cpu` / `libc10` symbols directly, the `libtorch_cuda` link request is silently dropped during the final link of the binary.
3. **`libtorch` performs lazy CUDA backend registration**: it only initializes the CUDA backend if `libtorch_cuda.so` (and `libc10_cuda.so`) is *loaded into the process address space*. There is no runtime `dlopen` fallback — `libtorch` checks the global symbol table on init. So a binary with no `DT_NEEDED libtorch_cuda.so` entry will return `false` from `tch::Cuda::is_available()` even on a CUDA-equipped box.

The runtime failure mode is *silent*: `Device::cuda_if_available()` returns `Device::Cpu` without warning, training proceeds on CPU at ~1/5th the expected throughput, and only careful benchmarking reveals it.

**Note on the existing `--no-as-needed` workaround**: `scripts/train-gpu.sh:59` and `scripts/gpu-train.sh:38` already set `RUSTFLAGS="-C link-arg=-Wl,--no-as-needed"`. In practice this is **insufficient by itself** — `docs/GPU_SETUP.md:36-40` documents that "the linker drops CUDA library as 'unused' even with `--no-as-needed`" and the `LD_PRELOAD` step is still required. The likely reason is that `RUSTFLAGS` `-C link-arg` adds the flag at a position in the linker command line that does not cover the `-ltorch_cuda` argument emitted by `torch-sys` (linker flag scope is positional: `--no-as-needed` only affects libraries *after* it on the command line, until `--as-needed` is seen). The fix requires the flag to be emitted *by the build script* immediately before `-ltorch_cuda`.

### Implementation Options

**Option A: Upstream fix in `torch-sys` build.rs (preferred, but external dependency)**

Patch `torch-sys` to emit positional `cargo:rustc-link-arg` directives that bracket the CUDA libs:

```rust
// In torch-sys/build.rs, when use_cuda is true:
println!("cargo:rustc-link-arg=-Wl,--no-as-needed");
println!("cargo:rustc-link-lib=torch_cuda");
println!("cargo:rustc-link-lib=c10_cuda");
println!("cargo:rustc-link-arg=-Wl,--as-needed");
```

This guarantees the CUDA libs survive `--as-needed` regardless of downstream RUSTFLAGS positioning. Requires a PR to https://github.com/LaurentMazare/tch-rs and a version bump. **Out of scope for this repo** in the short term.

**Option B: Force-reference a CUDA symbol from thrust's own code (in-repo, recommended)**

Add a `#[used]` static or a deliberate call to `tch::Cuda::is_available()` early in initialization, *inside a function the linker cannot eliminate*. The trick is that `tch::Cuda::is_available()` ultimately resolves to a symbol in `libtorch_cuda` only when CUDA support is compiled in; otherwise it stubs to false. Adding it in main and gating with the `training` feature defeats `--as-needed` for CUDA-enabled builds:

```rust
// src/lib.rs or a new src/cuda_anchor.rs (feature = "training" only)
#[cfg(feature = "training")]
#[allow(dead_code)]
#[doc(hidden)]
pub fn _cuda_link_anchor() -> bool {
    // Touching this symbol drags libtorch_cuda into the link.
    tch::Cuda::is_available()
}
```

Then call `crate::_cuda_link_anchor()` from a `static` initializer or a stub used by every `Device::cuda_if_available()` callsite. **Caveat**: `tch::Cuda::is_available()` may not pull a `libtorch_cuda`-only symbol; needs experimental verification (build, then `ldd target/.../bin | grep cuda` and `readelf -d target/.../bin | grep NEEDED`).

A more reliable variant is to bind a known CUDA-only `extern "C"` symbol via `extern "C" { fn at_cuda_synchronize(); }` (or similar from `c10::cuda` namespace) and reference its address — `#[used] static _ANCHOR: unsafe extern "C" fn() = at_cuda_synchronize;` — gated behind `cfg(feature = "training")` and possibly a new `cfg(thrust_cuda)` flag set by a thin `build.rs` that mirrors `torch-sys`'s CUDA detection.

**Option C: Custom build.rs emitting positional link args (in-repo, robust)**

Add a top-level `build.rs` to `thrust-rl` that:

1. Detects whether CUDA is enabled (mirror `torch-sys` detection logic, or read `cargo:rustc-cfg=use_cuda` cascaded from `torch-sys` if available).
2. Emits, when CUDA is on:
   ```
   cargo:rustc-link-arg=-Wl,--no-as-needed
   cargo:rustc-link-lib=torch_cuda
   cargo:rustc-link-lib=c10_cuda
   cargo:rustc-link-arg=-Wl,--as-needed
   ```

`cargo:rustc-link-arg` from a leaf-crate build.rs is appended *after* dependency link args, so the positional `--no-as-needed` will appear at the end of the link command — covering our explicit re-emission of `-ltorch_cuda`. This is the most surgical fix that doesn't depend on upstream `tch-rs`.

**Option D: Status quo + better detection (fallback)**

Keep the `LD_PRELOAD` workaround in scripts, but add a startup check that hard-errors if CUDA-flagged builds end up on CPU:

```rust
// In examples that expect GPU:
let device = Device::cuda_if_available();
#[cfg(feature = "training")]
if std::env::var("THRUST_EXPECT_CUDA").is_ok() && device == Device::Cpu {
    eprintln!("FATAL: THRUST_EXPECT_CUDA was set but tch fell back to Cpu.");
    eprintln!("Hint: ensure LD_PRELOAD includes libtorch_cuda.so + libc10_cuda.so.");
    std::process::exit(2);
}
```

Plus update `docs/GPU_SETUP.md` and `docs/GPU_TRAINING_GUIDE.md` to mark the `LD_PRELOAD` step as **REQUIRED**, not just helpful.

### Recommendation

Pursue **Option C** as the primary fix (in-repo `build.rs` with positional link args), with **Option D** as a hardening guard. Option C should make `cargo build` produce a binary that links `libtorch_cuda` correctly without any runtime `LD_PRELOAD` gymnastics. Option D ensures we fail loudly if Option C ever regresses or if the user is on a setup where it doesn't apply (e.g., a custom libtorch path).

Once Option C is verified working, the `LD_PRELOAD` exports in `scripts/gpu-train.sh:35`, `scripts/train-gpu.sh:55`, and `scripts/train-snake-remote.sh:28` can be removed (or kept as defensive belt-and-suspenders).

Defer Option A (upstream `torch-sys` patch) to a separate follow-up issue with the `tch-rs` maintainers.

### Acceptance Criteria

- [ ] `cargo build --release --example test_cuda` on a CUDA-enabled Linux box produces a binary where `ldd <binary> | grep torch_cuda` returns a non-empty line (and similarly for `libc10_cuda`).
- [ ] Running that binary **without** `LD_PRELOAD` set prints `tch::Cuda::is_available() = true` and `Device::cuda_if_available() = Cuda(0)`.
- [ ] `readelf -d target/release/examples/test_cuda | grep NEEDED` lists `libtorch_cuda.so` and `libc10_cuda.so`.
- [ ] CPU-only macOS builds (no CUDA) still build and run correctly; no spurious CUDA link errors on Darwin.
- [ ] The `LD_PRELOAD` workaround in `scripts/gpu-train.sh`, `scripts/train-gpu.sh`, and `scripts/train-snake-remote.sh` is either removed or annotated as "defensive only, no longer required by build".
- [ ] `docs/GPU_SETUP.md` and `docs/GPU_TRAINING_GUIDE.md` are updated to reflect the new build behavior (the "Key Discoveries" / "CUDA Detection Problem" sections need revising — the bug is no longer present).
- [ ] If Option D guard is added: a build with `THRUST_EXPECT_CUDA=1` on a box where CUDA libs failed to link produces a clear fatal error instead of silently running on CPU.

### Affected Files

- `build.rs` (new file at repo root) — implements Option C: detect CUDA, emit positional `-Wl,--no-as-needed` + `-ltorch_cuda -lc10_cuda` + `-Wl,--as-needed`. May need to mirror `torch-sys`'s CUDA detection logic, or rely on inspecting `DEP_TORCH_SYS_*` env vars set by torch-sys's build script for downstream crates.
- `Cargo.toml` (lines 1-16) — may need `build = "build.rs"` declaration and possibly a `[build-dependencies]` section if we need helpers (likely just `std::env`).
- `src/lib.rs` — (optional, Option B variant) add `#[used]` CUDA anchor symbol gated behind `cfg(feature = "training")`.
- `examples/test_cuda.rs` — extend to verify `readelf`/`ldd` output programmatically, or print a clear "CPU fallback detected" warning when expected to be on GPU.
- `scripts/gpu-train.sh:34-38` — remove or comment-out `LD_PRELOAD` line once Option C is verified.
- `scripts/train-gpu.sh:53-59` — same.
- `scripts/train-snake-remote.sh:27-29` — same.
- `docs/GPU_SETUP.md:22-63, 214-223` — rewrite "The CUDA Detection Problem" section; the bug is no longer required to be lived with.
- `docs/GPU_TRAINING_GUIDE.md:53-68` — same; LD_PRELOAD becomes optional/defensive rather than CRITICAL.

### Test Plan

- [ ] **Manual verification on Linux GPU box** (e.g. `rwalters-sandbox-2`):
  1. `git pull` the fix branch.
  2. `unset LD_PRELOAD; unset RUSTFLAGS`.
  3. `export LIBTORCH_USE_PYTORCH=1; export LIBTORCH_BYPASS_VERSION_CHECK=1`.
  4. `cargo clean && cargo build --release --example test_cuda`.
  5. `ldd target/release/examples/test_cuda | grep -E '(torch_cuda|c10_cuda)'` — must return both lines.
  6. `./target/release/examples/test_cuda` — must print `Cuda(0)` and `tch::Cuda::is_available() = true`.
- [ ] **Manual verification on macOS** (CPU-only):
  1. `cargo build --release --example test_cuda` succeeds.
  2. Binary runs without CUDA-related dyld errors.
  3. Reports `Device::Cpu` correctly.
- [ ] **Regression test for training throughput**:
  1. Run `./scripts/gpu-train.sh train_cartpole_long` *without* the `LD_PRELOAD` line in the script.
  2. Confirm `nvidia-smi -l 1` shows GPU utilization > 50% during training.
  3. Confirm training throughput matches or exceeds previous `LD_PRELOAD`-based runs (no regression).
- [ ] **Build-script idempotency**: `cargo build` twice in a row should not re-run the new `build.rs` unnecessarily (proper `cargo:rerun-if-changed=` directives).
- [ ] **Automated tests**: N/A for link-time behavior. Could add a CI step that runs `ldd | grep torch_cuda` on the test_cuda binary if a GPU runner is available.
- [ ] **Edge cases**:
  - CPU-only feature builds (`--no-default-features --features wasm`) must not pull in CUDA link args.
  - `LIBTORCH_USE_PYTORCH=0` with a system libtorch that has no CUDA must not break the build.
  - Different PyTorch versions (2.9.x patch versions) should all work.

### Dependencies

None. Issue #4 and #8 are tangentially related (same "tch is brittle" family) but do not block this fix. This issue can proceed independently.

### Out of Scope

- Upstream patch to `torch-sys` / `tch-rs` (file as a separate issue with maintainers if Option C proves insufficient).
- Windows support (`LD_PRELOAD` equivalent is `PATH` DLL ordering, different bug class).
- macOS Metal/MPS device detection (not affected by Linux `--as-needed`).
- Refactoring the GPU training script suite (`gpu-train.sh` vs `train-gpu.sh` vs `train-snake-remote.sh` duplication — separate cleanup issue).

**Comment by rjwalters** [comment]
Curator: Enhanced this issue per user request with root-cause analysis, four implementation options (A: upstream torch-sys fix, B: in-repo symbol anchor, C: in-repo build.rs with positional link args [recommended], D: status-quo + hard-fail guard), explicit acceptance criteria, an Affected Files inventory, and a verification-focused Test Plan.

Key insight surfaced during curation: the existing `RUSTFLAGS="-C link-arg=-Wl,--no-as-needed"` workaround (in `scripts/gpu-train.sh:38`, `scripts/train-gpu.sh:59`, `scripts/train-snake-remote.sh:29`) is **positionally ineffective** against the `-ltorch_cuda` emitted by `torch-sys`, which is why `LD_PRELOAD` is still required even though we already set `--no-as-needed`. `docs/GPU_SETUP.md:36-40` corroborates this. The recommended fix (Option C, a leaf-crate `build.rs`) re-emits the link args at a position where they actually cover `-ltorch_cuda`.

Marking `loom:curated`. Awaiting human approval via `loom:issue` label before Builder claims.
